File: learn/graph/lca/lca.go


type LCA struct {
	up    [][]int
	depth []int
	log   int
}

func NewLCA(n int, log int) *LCA {
	up := make([][]int, n)
	for i := range up {
		up[i] = make([]int, log)
	}
	return &LCA{
		up:    up,
		depth: make([]int, n),
		log:   log,
	}
}

func (lca *LCA) Dfs(v int, p int, adjList [][]int) {
	lca.up[v][0] = p
	if p != v {
		lca.depth[v] = lca.depth[p] + 1
	}

	for i := 1; i < lca.log; i++ {
		lca.up[v][i] = lca.up[lca.up[v][i-1]][i-1]
	}

	for _, u := range adjList[v] {
		if u != p {
			lca.Dfs(u, v, adjList)
		}
	}
}

func (lca *LCA) GetLCA(u, v int) int {
	if lca.depth[u] < lca.depth[v] {
		u, v = v, u
	}

	k := lca.depth[u] - lca.depth[v]
	for i := lca.log - 1; i >= 0; i-- {
		if (k & (1 << i)) != 0 {
			u = lca.up[u][i]
		}
	}

	if u == v {
		return u
	}

	for i := lca.log - 1; i >= 0; i-- {
		if lca.up[u][i] != lca.up[v][i] {
			u = lca.up[u][i]
			v = lca.up[v][i]
		}
	}

	return lca.up[u][0]
}

File: learn/graph/lca/lca.rs
pub struct LCA {
    up: Vec<Vec<usize>>,
    depth: Vec<usize>,
    log: usize,
}

impl LCA {
    pub fn new(n: usize, log: usize) -> Self {
        LCA {
            up: vec![vec![0; log]; n],
            depth: vec![0; n],
            log,
        }
    }

    pub fn dfs(&mut self, v: usize, p: usize, adj_list: &Vec<Vec<usize>>) {
        self.up[v][0] = p;
        if p != v {
            self.depth[v] = self.depth[p] + 1;
        }

        for i in 1..self.log {
            self.up[v][i] = self.up[self.up[v][i - 1]][i - 1];
        }

        for &u in &adj_list[v] {
            if u != p {
                self.dfs(u, v, adj_list);
            }
        }
    }

    pub fn get_lca(&self, mut u: usize, mut v: usize) -> usize {
        if self.depth[u] < self.depth[v] {
            std::mem::swap(&mut u, &mut v);
        }

        let k = self.depth[u] - self.depth[v];
        for i in (0..self.log).rev() {
            if (k & (1 << i)) != 0 {
                u = self.up[u][i];
            }
        }

        if u == v {
            return u;
        }

        for i in (0..self.log).rev() {
            if self.up[u][i] != self.up[v][i] {
                u = self.up[u][i];
                v = self.up[v][i];
            }
        }

        self.up[u][0]
    }
}
